File: 2021-22/as2122_5a_sia/2022-01-11/radq.php
<?php
// Metodo di Erone per il calcolo della radice quadrata
function radq($x) {
  return radq_iter($x, 1.0);
}

function radq_iter($x, $stima) {
  if (suff_accurata($stima, $x))
    return $stima;
  return radq_iter($x, migliora_stima($stima, $x));
}

function suff_accurata($stima, $x) {
  return valore_assoluto(quadrato($stima) - $x) <= 1E-9;
}

function valore_assoluto($x) {
  if ($x >= 0) return $x;
  return -$x;
}

function quadrato($x) {
  return $x * $x;
}

function migliora_stima($stima, $x) {
  return valor_medio($stima, $x / $stima);
}

function valor_medio($x, $y) {
  return ($x + $y) / 2;
}

echo radq(4) . "\n";
echo radq(9) . "\n";
echo radq(16) . "\n";
echo radq(2) . "\n";
echo radq(3) . "\n";
?>

File: 2021-22/as2122_3asia/2022-01-12/radq.cpp
#include <iostream>
using namespace std;

// Prototipi
double radq(double);
double radq_iterazione(double, double);
bool sufficientemente_accurata(double, double, double);
double valore_assoluto(double);
double quadrato(double);
double migliora_stima(double, double);
double valor_medio(double, double);

int main() {
  double x;
  cout << "Immetti il numero di cui estrarre la radice quadrata: ";
  cin >> x;
  cout << "radq(" << x << ") = " << radq(x) << endl;

  return 0;
}

/** Estrae la radice quadrata del radicando
 *
 * @param radicando Il radicando
 * @return un'apporossimazione della radice quadrata
 */
double radq(double radicando) {
  return radq_iterazione(radicando, 1.0);
} 

/** Esegue il processo di approssimazione della radice quadrata
 *
 * @param radicando Il radicando
 * @param stima Una stima della radice quadrata
 * @return La migliore stima della radice quadrata
 */
double radq_iterazione(double radicando, double stima) {
  return (sufficientemente_accurata(stima, radicando, 5E-7))
          ? stima
          : radq_iterazione(radicando, migliora_stima(stima, radicando));
}

/** Determina se la stima è sufficientemente accurata
 *
 * @param stima La stima
 * @param radicando Il radicando
 * @param tolleranza La tolleranza che determina la qualità della stima
 * @return vero se la stima è sufficientemente accurata nel rispetto della tolleranza
 */
bool sufficientemente_accurata(double stima, double radicando, double tolleranza) {
  return valore_assoluto(
            quadrato(stima) - radicando
         ) <= tolleranza;
}

/** Valore assoluto, detto anche modulo, di un numero reale
 *
 * @param x
 * @return |x|
 */
double valore_assoluto(double x) {
  return (x >= 0)
          ? x
          : -x;
}

/** Restituisce la seconda potenza di un numero
 *
 * @param x Il numero
 * @return $x^2$
 */
double quadrato(double x) {
  return x * x;
}

/** Migliora la stima della radice quadrata del radicando
 * 
 * @param stima La stima
 * @param radicando Il radicando
 * @return una nuova stima
 */
double migliora_stima(double stima, double radicando) {
  return valor_medio(stima, radicando / stima);
}

/** Determina il valore medio tra due numeri
 *
 * @param x Il primo numero
 * @parm y Il secondo numero
 * @return Il valor medio tra x e y
 */
double valor_medio(double x, double y) {
  return (x + y) / 2;
}
